include/boost/random/faure.hpp

```diff
@@ -0,0 +1,342 @@
+
+      *it = compute_recip(seq, hisum, ytemp.rbegin());
+
+      // Find other components using the Faure method.
+      ++it;
+      for ( ; it != quasi.end(); ++it)
+      {
+        *it = RealType();
+        RealType r = inv_qs_base;
+
+        for (std::size_t i = 0; i != hisum; ++i)
+        {
+          RealType ztemp = RealType();
+          for (std::size_t j = i; j != hisum; ++j)
+            ztemp += ytemp[j] * upper_element(i, j, hisum);
+
+          // Sum ( J <= I <= HISUM ) ( old ytemp(i) * binom(i,j) ) mod QS.
+          ytemp[i] = std::fmod(ztemp, static_cast<RealType>(qs_base));
+          *it += ytemp[i] * r;
+          r *= inv_qs_base;
+        }
+      }
+    }
+  }
+
+private:
+  inline std::size_t n_elements(std::size_t seq) const
+  {
+    return integer_log(qs_base, seq) + 1;
+  }
+
+  inline static std::size_t size_hint(std::size_t n)
+  {
+    return n * (n + 1) / 2;
+  }
+
+  packed_uint_t& upper_element(std::size_t i, std::size_t j, std::size_t dim)
+  {
+    BOOST_ASSERT( i < dim );
+    BOOST_ASSERT( j < dim );
+    BOOST_ASSERT( i <= j );
+    return coeff[(i * (2 * dim - i + 1)) / 2 + j - i];
+  }
+
+  template<typename Iterator>
+  RealType compute_recip(std::size_t seq, std::size_t n, Iterator out) const
+  {
+    // Here we do
+    //   Sum ( 0 <= J <= HISUM ) YTEMP(J) * QS**J
+    //   Sum ( 0 <= J <= HISUM ) YTEMP(J) / QS**(J+1)
+    // in one go
+    RealType r = RealType();
+    std::size_t m, k = integer_pow(qs_base, n - 1);
+    for( ; n != 0; --n, ++out, seq = m, k /= qs_base )
+    {
+      m  = seq % k;
+      RealType v  = (seq - m) / k; // RealType <- IntType
+      r += v;
+      r *= inv_qs_base;
+      *out = v; // saves double dereference
+    }
+    return r;
+  }
+
+  void compute_coefficients(const std::size_t n)
+  {
+    // Resize and initialize to zero
+    coeff.resize(size_hint(n));
+    std::fill(coeff.begin(), coeff.end(), packed_uint_t());
+
+    // The first row and the diagonal is assigned to 1
+    upper_element(0, 0, n) = 1;
+    for (std::size_t i = 1; i < n; ++i)
+    {
+      upper_element(0, i, n) = 1;
+      upper_element(i, i, n) = 1;
+    }
+
+    // Computes binomial coefficients MOD qs_base
+    for (std::size_t i = 1; i < n; ++i)
+    {
+      for (std::size_t j = i + 1; j < n; ++j)
+      {
+        upper_element(i, j, n) = ( upper_element(i, j-1, n) +
+                                   upper_element(i-1, j-1, n) ) % qs_base;
+      }
+    }
+  }
+
+  void recompute_tables(std::size_t n)
+  {
+    ytemp.resize(n);
+    compute_coefficients(n);
+  }
+
+private:
+  packed_uint_t qs_base;
+  RealType inv_qs_base;
+
+  // here we cache precomputed data; note that binomial coefficients have
+  // to be recomputed iff the integer part of the logarithm of seq changes,
+  // which happens relatively rarely.
+  std::vector<packed_uint_t> coeff; // packed upper (!) triangular matrix
+  std::vector<RealType> ytemp;
```

Comment:
Does @swatanabe have a convention for member variables to follow, like `m_` prefixes?